Enable jit tracing to parametrization and add jit tests

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60969 Enable jit tracing to parametrization and add jit tests**

This PR fixes the tracing in the parametrizations.
The current resolution is that when tracing is performed while caching is enabled, we throw an error.
Without caching, the tracing should work properly (tests added).

Currently, the parametrizations don't support scripting.
This PR introduces the same logic as with the tracing (throw error if caching).
However, the scripting itself cannot enabled due to the use of the generator expressions in the parametrizations.
Added TODO to fix it.

Differential Revision: [D29462887](https://our.internmc.facebook.com/intern/diff/D29462887)

Commit:
Enable jit tracing to parametrization and add jit tests

This PR fixes the tracing in the parametrizations.
The current resolution is that when tracing is performed while caching is enabled, we throw an error.
Without caching, the tracing should work properly (tests added).

Currently, the parametrizations don't support scripting.
This PR introduces the same logic as with the tracing (throw error if caching).
However, the scripting itself cannot enabled due to the use of the generator expressions in the parametrizations.
Added TODO to fix it.

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29462887).

Commit:
Update on "Enable jit tracing to parametrization and add jit tests"


This PR fixes the tracing in the parametrizations.
The current resolution is that when tracing is performed while caching is enabled, we throw an error.
Without caching, the tracing should work properly (tests added).

Currently, the parametrizations don't support scripting.
This PR introduces the same logic as with the tracing (throw error if caching).
However, the scripting itself cannot enabled due to the use of the generator expressions in the parametrizations.
Added TODO to fix it.

Differential Revision: [D29462887](https://our.internmc.facebook.com/intern/diff/D29462887)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29462887).

albanD commented on test/test_nn.py:
```diff
@@ -2782,6 +2782,43 @@ def forward(self, X):
             Y = model.weight
             self.assertEqual(id(X), id(Y))
 
+    def test_jit_parametrization(self):
```
Do these usually live in test_nn or test_jit?

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -300,17 +300,31 @@ def _inject_property(module: Module, tensor_name: str) -> None:
+            elif torch._C._get_tracing_state() is not None:
+                # Tracing
+                raise RuntimeError('Caching should be disabled while tracing a '
+                                   'model. You can run a traced model with '
```
The caching flag will actually be ignored while running the traced model no?
If so, I think this message is a bit misleading as I feel like it implies that the flag will be honored when running the traced model.

z-a-f replied:
The caching flag might be enabled during the tracing. This line is here to catch these types of implementations::

```python
with parametrize.cached():
    traced_model = torch.jit.trace_module(model, {'forward': x})
```

I think the message could be improved. How about

```suggestion
                                   'Cannot trace a model while caching parametrizations.'
```

Alternatively, we don't throw an error here, and quietly just do 

```suggestion
                                   return parametrization()
```

albanD replied:
No, the error is good. And the updated error message is good as well I think!

albanD reviewed (commented):
Code change looks good.
Just small details about message and test location.

Commit:
Update on "Enable jit tracing to parametrization and add jit tests"


This PR fixes the tracing in the parametrizations.
The current resolution is that when tracing is performed while caching is enabled, we throw an error.
Without caching, the tracing should work properly (tests added).

Currently, the parametrizations don't support scripting.
This PR introduces the same logic as with the tracing (throw error if caching).
However, the scripting itself cannot enabled due to the use of the generator expressions in the parametrizations.
Added TODO to fix it.

Differential Revision: [D29462887](https://our.internmc.facebook.com/intern/diff/D29462887)

[ghstack-poisoned]

albanD reviewed (approved):
Looks good!
Thanks for the update

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29462887).
